internal/feature/features.go

```diff
@@ -0,0 +1,126 @@
+
+# Using Features
+
+We initialize and configure one [MutableGate] in main() and add it to the Context
+passed to Reconcilers and other Runnables. Those can then interrogate it using [Enabled]:
+
+	if !feature.Enabled(ctx, feature.Excellent) { return }
+
+Tests should create and configure their own [MutableGate] and inject it using
+[NewContext]. For example, the following enables one feature and disables another:
+
+	gate := feature.NewGate()
+	assert.NilError(t, gate.SetFromMap(map[string]bool{
+		feature.Excellent: true,
+		feature.Uncommon: false,
+	}))
+	ctx := feature.NewContext(context.Background(), gate)
+*/
+package feature
+
+import (
+	"context"
+
+	"k8s.io/component-base/featuregate"
+)
+
+type Feature = featuregate.Feature
+
+// Gate indicates what features exist and which are enabled.
+type Gate interface {
+	Enabled(Feature) bool
+	String() string
+}
+
+// MutableGate contains features that can be enabled or disabled.
+type MutableGate interface {
+	Gate
+	// Set enables or disables features by parsing a string like "feature1=true,feature2=false".
+	Set(string) error
+	// SetFromMap enables or disables features by boolean values.
+	SetFromMap(map[string]bool) error
+}
+
+const (
+	// Support appending custom queries to default PGMonitor queries
+	AppendCustomQueries = "AppendCustomQueries"
+
+	// Support automatically growing volumes
+	AutoGrowVolumes = "AutoGrowVolumes"
+
+	BridgeIdentifiers = "BridgeIdentifiers"
+
+	// Support custom sidecars for PostgreSQL instance Pods
+	InstanceSidecars = "InstanceSidecars"
+
+	// Support custom sidecars for pgBouncer Pods
+	PGBouncerSidecars = "PGBouncerSidecars"
+
+	// Support tablespace volumes
+	TablespaceVolumes = "TablespaceVolumes"
+)
+
+// NewGate returns a MutableGate with the Features defined in this package.
+func NewGate() MutableGate {
+	gate := featuregate.NewFeatureGate()
+
+	if err := gate.Add(map[Feature]featuregate.FeatureSpec{
+		AppendCustomQueries: {Default: false, PreRelease: featuregate.Alpha},
+		AutoGrowVolumes:     {Default: false, PreRelease: featuregate.Alpha},
+		BridgeIdentifiers:   {Default: false, PreRelease: featuregate.Alpha},
+		InstanceSidecars:    {Default: false, PreRelease: featuregate.Alpha},
+		PGBouncerSidecars:   {Default: false, PreRelease: featuregate.Alpha},
+		TablespaceVolumes:   {Default: false, PreRelease: featuregate.Alpha},
+	}); err != nil {
+		panic(err)
+	}
+
+	return gate
+}
+
+type contextKey struct{}
+
+// Enabled indicates if a Feature is enabled in the Gate contained in ctx. It
+// returns false when there is no Gate.
+func Enabled(ctx context.Context, f Feature) bool {
+	gate, ok := ctx.Value(contextKey{}).(Gate)
+	return ok && gate.Enabled(f)
+}
+
+// NewContext returns a copy of ctx containing gate. Check it using [Enabled].
+func NewContext(ctx context.Context, gate Gate) context.Context {
+	return context.WithValue(ctx, contextKey{}, gate)
```

Comment:
Why `contextKey{}` as the key?

Reply:
The key is passed and compared as an interface, so this unexported type gives a unique value that also requires fewer (no?) allocations.

- https://pkg.go.dev/context#WithValue
- https://go.dev/ref/spec#Comparison_operators<br><br>
  > Two interface values are equal if they have [identical](https://go.dev/ref/spec#Type_identity) dynamic types and equal dynamic values or if both have value `nil`.
- https://dave.cheney.net/2014/03/25/the-empty-struct